asr-diarization on IE blog (#1991)
huggingface/blog: asr-diarization.md

Reviewer: We also need to add a metadata entry to `_blog.yml`

Author: haven't figured out the thumbnail part yet :)

---

> title: "ASR+Diarization on Inference Endpoints" 

Reviewer 2: Change title here as well

thumbnail: /blog/assets/asr-diarization/thumbnail.png
authors:
- user: sergeipetrov
- user: reach-vb
---

> # Transcribe Speech and Speaker Diarise on Hugging Face Inference Endpoints

Reviewer 2: Maybe we can make it a bit more catchy? 

"Fastes way to transcribe audio...."

Author: @Vaibhavs10 would it be technically correct?

Reviewer: (nit: consistent spelling of diarization / diarisation)

Reviewer: How about `Powerful ASR + diarization + speculative decoding with Hugging Face Inference Endpoints`? A bit too long maybe?

Author: It's most descriptive of what it's all about so changed to this!

> Whisper is one of the best open source speech recognition models and definitely the one most widely used. Hugging Face [Inference Endpoints](https://huggingface.co/inference-endpoints/dedicated) make it very easy to deploy it out-of-the-box. However, if you’d like to couple it with a diarization pipeline to identify speakers or introduce assisted generation for speculative decoding, things may get tricky. 

Reviewer: ```suggestion
Whisper is one of the best open source speech recognition models and definitely the one most widely used. Hugging Face [Inference Endpoints](https://huggingface.co/inference-endpoints/dedicated) make it very easy to deploy any Whisper model out of the box. However, if you’d like to
introduce additional features, like a diarization pipeline to identify speakers, or assisted generation for speculative decoding, things get trickier. The reason is that you need to combine Whisper with additional models, while still exposing a single API endpoint.
```

> In this blog, we’ll explore how to deploy the Automatic Speech Recogniton (ASR) and Diarization pipeline on Inference Endpoints. We’ll focus on how to do so using a [custom inference handler](https://huggingface.co/docs/inference-endpoints/guides/custom_handler). The implementation of the pipeline is inspired by the famous [Insanely Fast Whisper](https://github.com/Vaibhavs10/insanely-fast-whisper#insanely-fast-whisper) and uses a [Pyannote](https://github.com/pyannote/pyannote-audio) model for diarization. For those ready to go all the way, we provide a containerized model server that can be deployed anywhere. 

Reviewer: ```suggestion
We'll solve this challenge using a [custom inference handler](https://huggingface.co/docs/inference-endpoints/guides/custom_handler), which will implement the Automatic Speech Recogniton (ASR) and Diarization pipeline on Inference Endpoints, as well as supporting speculative decoding. The implementation of the diarization pipeline is inspired by the famous [Insanely Fast Whisper](https://github.com/Vaibhavs10/insanely-fast-whisper#insanely-fast-whisper), and it uses a [Pyannote](https://github.com/pyannote/pyannote-audio) model for diarization. For those ready to go all the way, we'll also provide a containerized model server that can be deployed anywhere. 
```

This will also be a demonstration of how flexible Inference Endpoints are and that you can host pretty much anything there. Here is the code to follow along:

- [A custom ASR and Diarization handler](https://huggingface.co/sergeipetrov/asrdiarization-handler-default/blob/main/handler.py) - this handler works on Inference Endpoints as-is.

> - [A standalone platform-agnostic model server](https://github.com/plaggy/fast-whisper-server/tree/main/model-server) - this is a container that you can deploy anywhere - including Inference Endpoints! This is also a good fit if you’d like to do something very custom.

Reviewer: Is this from someone in the community? If so, a shout out to them would be nice.

At this point through the first reading I'm not sure what the container is for, will get back to it later.

Reviewer: Edit: I see it's you :)

> **_Starting with [Pytorch 2.2](https://pytorch.org/blog/pytorch2-2/) SDPA supports Flash Attention 2 out-of-the-box so you don’t need to install and pass it explicitly._**

Reviewer: ```suggestion
**_Starting with [Pytorch 2.2](https://pytorch.org/blog/pytorch2-2/), SDPA supports Flash Attention 2 out-of-the-box, so we'll use that version for faster inference._**
```

> ### The main modules

Reviewer: ```suggestion
## The main modules
```

> Below is a high-level schema of what the endpoint looks like under the hood:

Reviewer: ```suggestion
This is a high-level diagram of what the endpoint looks like under the hood:
```

![pipeline_schema](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/asr-diarization/pipeline_schema.png)

> The implementation of ASR and diarization pipelines is modularized to cater to a wider range of use cases - the diarization pipeline operates on top of ASR outputs, and you can use only the ASR part if diarization is not needed. For diarization, we propose using the [Pyannote model](https://huggingface.co/pyannote/speaker-diarization-3.1), which is a current open source SOTA.

Reviewer: ```suggestion
The implementation of ASR and diarization pipelines is modularized to cater to a wider range of use cases - the diarization pipeline operates on top of ASR outputs, and you can use only the ASR part if diarization is not needed. For diarization, we propose using the [Pyannote model](https://huggingface.co/pyannote/speaker-diarization-3.1), currently a SOTA open source implementation.
```

> We’ll add speculative decoding as a way to speed up inference. The speedup is achieved by using a smaller and faster model to do expensive auto-regressive generation. Learn more about how it works with Whisper specifically in [this](https://huggingface.co/blog/whisper-speculative-decoding) great blog post.

Reviewer: ```suggestion
We’ll also add speculative decoding as a way to speed up inference. The speedup is achieved by using a smaller and faster model to suggest generations that are validated by the larger model. Learn more about how it works with Whisper specifically in [this great blog post](https://huggingface.co/blog/whisper-speculative-decoding).
```

Speculative decoding comes with restrictions:

> - at least the decoder part of an assistant model should be the same as that of the main model

Reviewer: ```suggestion
- at least the decoder part of an assistant model should be the same as that of the main model
```

The same? Or the same architecture?

- the batch size much be 1

> Make sure to take the above into account: at some point you wouldn’t want to trade a larger batch for an assistant model. A great choice for an assistant model for Whisper is a corresponding [distilled version](https://huggingface.co/distil-whisper).

Reviewer: ```suggestion
Make sure to take the above into account. Depending on your production use case, supporting larger batches can be faster than speculative decoding.

If you do use an assistant model, a great choice for Whisper is a [distilled version](https://huggingface.co/distil-whisper).
```

Maybe also anticipate here that you can skip the assistant model by setting this property to `None`: https://huggingface.co/sergeipetrov/asrdiarization-handler/blob/main/config.py#L12, even if ModelSettings is explained later.

> ### Set up your own

Reviewer: ```suggestion
## Set up your own endpoint
```

> The easiest way to start is to clone the custom handler repository using the [repo duplicator](https://huggingface.co/spaces/huggingface-projects/repo_duplicator).

Reviewer 3: We should mention which repo are we talking about here -> we can add you repo to `hf-audio` org if you want too.

Author: I added the reference but what you suggested is cooler ofc

Here is the model loading piece from the `handler.py`:

```python
from pyannote.audio import Pipeline
from transformers import pipeline, AutoModelForCausalLM

...

self.asr_pipeline = pipeline(
      "automatic-speech-recognition",
      model=model_settings.asr_model,
      torch_dtype=torch_dtype,
      device=device
  )

  self.assistant_model = AutoModelForCausalLM.from_pretrained(
      model_settings.assistant_model,
      torch_dtype=torch_dtype,
      low_cpu_mem_usage=True,
      use_safetensors=True
  ) 
  
  ...

  self.diarization_pipeline = Pipeline.from_pretrained(
      checkpoint_path=model_settings.diarization_model,
      use_auth_token=model_settings.hf_token,
  ) 
  
  ...
```

> You can customize the pipeline based on your needs. `ModelSettings` in the `config.py` hold the parameters used at the initialization time that define which models to use:

Reviewer: ```suggestion
You can customize the pipeline based on your needs. `ModelSettings`, in the `config.py` file, holds the parameters used for initialization, defining the models to use during inference:
```

```python
class ModelSettings(BaseSettings):
    asr_model: str
    assistant_model: Optional[str] = None
    diarization_model: Optional[str] = None
    hf_token: Optional[str] = None

> ```

Reviewer 2: thats a bit confusing to me knowing that `handler.py` has one class

Author: Could you elaborate? A custom handler is in a repo, and the entire repo gets mounted (I think you implemented it). There may be multiple files and multiple classes in a repo talking to each other

Reviewer: Yes, but the documentation linked before only talks about a `handler.py` in the model repo that needs to implement a couple of entry points. Maybe explaining that the custom handler implementation can import other scripts is not out of the question. An alternative would be to provide a single-file implementation if you think that'd be easier to understand.

Author: A single-file would be very hard to get through, added the explanation of the mounting process and file descriptions in the beginning (3rd paragraph)

> The parameters can be adjusted by passing environment variables with corresponding names - this works both with a custom container and an inference handler. It’s a [Pydantic feature](https://docs.pydantic.dev/latest/concepts/pydantic_settings/). To pass environment variables to a container during build time you’ll have to create an endpoint via an API call (not via the interface). 

Reviewer: I'm a bit lost here with respect to the container vs handler discussion and the use of pydantic.

Author: It's all a handler now, lmk if it's still unclear

> You could hardcode model names not to pass them as environment variables but *note that the diarization pipeline requires a token to be passed explicitly (`hf_token`).*

Reviewer: ```suggestion
You could hardcode model names instead of passing them as environment variables, but *note that the diarization pipeline requires a token to be passed explicitly (`hf_token`).* This is better suited for an environment variable.
```

For context, all the diarization-related pre- and postprocessing utils are in `diarization_utils.py`

The only required component is an ASR model. Optionally, an assistant model can be specified to be used for speculative decoding, and a diarization model can be used to partition a transcription by speakers.

### Deploy on Inference Endpoints

If you only need the ASR part you could specify `asr_model`/`assistant_model` in the `config.py` and deploy with a click of a button:

![deploy_oneclick](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/asr-diarization/deploy_oneclick.png)

To pass environment variables to containers hosted on Inference Endpoints you’ll need to create an endpoint programmatically using the [provided API](https://api.endpoints.huggingface.cloud/#post-/v2/endpoint/-namespace-). Below is an example call:

```python

> body = {

Reviewer 2: maybe convert to a huggingface_hub example on how to create?

Author: Do you have an example? I don't know another way of doing it

    "compute": {
        "accelerator": "gpu",
        "instanceSize": "medium",
        "instanceType": "g5.2xlarge",
        "scaling": {
            "maxReplica": 1,
            "minReplica": 0
        }
    },
    "model": {
        "framework": "pytorch",
        "image": {
            # a default container
            "huggingface": {
                "env": {
		                # this is where a Hub model gets mounted
                    "HF_MODEL_DIR": "/repository", 
                    "DIARIZATION_MODEL": "pyannote/speaker-diarization-3.1",
                    "HF_TOKEN": "<your_token>",
                    "ASR_MODEL": "openai/whisper-large-v3",
                    "ASSISTANT_MODEL": "distil-whisper/distil-large-v3"
                }
            }
        },
        # a model repository on the Hub
        "repository": "sergeipetrov/asrdiarization-handler",
        "task": "custom"
    },
    # the endpoint name
    "name": "asr-diarization-1",
    "provider": {
        "region": "us-east-1",
        "vendor": "aws"
    },
    "type": "private"
}
```

### When to use an assistant model

> To give a better idea on when using an assistant model is beneficial given the speculative decoding restrictions below is benchmarking performed with [k6](https://k6.io/docs/):

Reviewer: ```suggestion
To give a better idea on when using an assistant model is beneficial, here's a benchmark performed with [k6](https://k6.io/docs/):
```

```bash
# Setup:
# GPU: A10
ASR_MODEL=openai/whisper-large-v3
ASSISTANT_MODEL=distil-whisper/distil-large-v3

# long: 60s audio; short: 8s audio
long_assisted..................: avg=4.15s    min=3.84s    med=3.95s    max=6.88s    p(90)=4.03s    p(95)=4.89s   
long_not_assisted..............: avg=3.48s    min=3.42s    med=3.46s    max=3.71s    p(90)=3.56s    p(95)=3.61s   
short_assisted.................: avg=326.96ms min=313.01ms med=319.41ms max=960.75ms p(90)=325.55ms p(95)=326.07ms
short_not_assisted.............: avg=784.35ms min=736.55ms med=747.67ms max=2s       p(90)=772.9ms  p(95)=774.1ms
```

> As you can see, assisted generation gives dramatic performance gains when an audio is short (batch size is 1). If an audio is longer than one chunk speculative decoding may hurt inference time because of the batch size restriction.

Reviewer: ```suggestion
As you can see, assisted generation gives dramatic performance gains when an audio is short (batch size is 1). If an audio is long, inference will automatically chunk it into batches, and speculative decoding may hurt inference time because of the limitations we discussed before.
```

(Not everybody may be familiar with how this works. When I was reading the intro, I thought that "batching" referred to the number of inference requests supported in parallel).

### Payload

> Once deployed, send your audio along with parameters your Inference Endpoints like this (in Python):

Reviewer: ```suggestion
Once deployed, send your audio along with the inference parameters to your inference endpoint, like this (in Python):
```

```python
import base64
import requests

API_URL = "<your endpoint URL>"
filepath = "/path/to/audio"

with open(filepath, "rb") as f:
    audio_encoded = base64.b64encode(f.read()).decode("utf-8")

data = {
    "inputs": audio_encoded,
    "parameters": {
        "batch_size": 24
    }
}

resp = requests.post(API_URL, json=data, headers={"Authorization": "Bearer <your token>"})
print(resp.json())
```

Or with [InferenceClient](https://huggingface.co/docs/huggingface_hub/en/package_reference/inference_client#huggingface_hub.InferenceClient) (there is also an [async version](https://huggingface.co/docs/huggingface_hub/en/package_reference/inference_client#huggingface_hub.AsyncInferenceClient)):

```python
from huggingface_hub import InferenceClient

client = InferenceClient(model = "<your endpoint URL>", token="<your token>")

with open("/path/to/audio", "rb") as f:
    audio_encoded = base64.b64encode(f.read()).decode("utf-8")
data = {
    "inputs": audio_encoded,
    "parameters": {
        "batch_size": 24
    }
}

res = client.post(json=data)
```

> ### Standalone model server

Reviewer: ```suggestion
## Standalone model server
```

> [A standalone containerized server](https://github.com/plaggy/fast-whisper-server/tree/main/model-server) has everything to turn a configured pipeline into an endpoint. It can be deployed with hosting services or on unmanaged infrastructure, depending on your needs.

Reviewer 2: if thats not used in the example for IE i would remove it

Reviewer: I tend to agree, some parts of the discussion could be simplified. A trade-off would be to just mention the container at the end of the article, for additional reference.

Author: it was an artifact of the initial implementation, when a custom container was needed to make flash attention 2 available. Removed completely

Author: Mentioned it at the very bottom. Let me know if it looks out of place

> **Recap**   

Reviewer: ```suggestion
## Recap
```

To recap, these are the options we made available for hosting the ASR+Diarization pipeline on Inference Endpoints:

- A custom model handler on the Hub. You could hardcode models to use or parametrize with environment variables. Remember that a Pyannote diarization pipeline requires a token so you’ll need to pass it as an environment variable to perform diarization.

> - A platform-agnostic containerized model service. Can be hosted on Inference Endpoints and anywhere else.

Reviewer: To be reassessed if there are any changes in the main content :)

    "provider": {
        "region": "us-east-1",
        "vendor": "aws"
    },
    "type": "private"
}
```

### When to use an assistant model

To give a better idea on when using an assistant model is beneficial, here's a benchmark performed with [k6](https://k6.io/docs/):

```bash
# Setup:
# GPU: A10
ASR_MODEL=openai/whisper-large-v3
ASSISTANT_MODEL=distil-whisper/distil-large-v3

# long: 60s audio; short: 8s audio
long_assisted..................: avg=4.15s    min=3.84s    med=3.95s    max=6.88s    p(90)=4.03s    p(95)=4.89s   
long_not_assisted..............: avg=3.48s    min=3.42s    med=3.46s    max=3.71s    p(90)=3.56s    p(95)=3.61s   
short_assisted.................: avg=326.96ms min=313.01ms med=319.41ms max=960.75ms p(90)=325.55ms p(95)=326.07ms
short_not_assisted.............: avg=784.35ms min=736.55ms med=747.67ms max=2s       p(90)=772.9ms  p(95)=774.1ms
```

As you can see, assisted generation gives dramatic performance gains when an audio is short (batch size is 1). If an audio is long, inference will automatically chunk it into batches, and speculative decoding may hurt inference time because of the limitations we discussed before.

### Inference parameters

> All the things you can parametrize on inference are in the `config.py`:

Reviewer: ```suggestion

All the inference parameters are in `config.py`:

```
